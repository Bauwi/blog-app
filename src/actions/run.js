import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  console.log(post);
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  console.log(DBid);
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      const run = [];
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      dispatch(setRunPosts(run));
    });
};
