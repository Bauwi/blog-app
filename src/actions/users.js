import { db } from '../firebase/firebase';

export const updateUser = (id, updates) => ({
  type: 'UPDATE_USER',
  id,
  updates
});

export const startUpdateUser = (id, updates) => (dispatch, getState) => {
  const uid = id || getState().auth.uid;
  return db
    .collection('users')
    .doc(uid)
    .get()
    .then((doc) => {
      if (doc.exists) {
        return db
          .collection('users')
          .doc(uid)
          .update(updates)
          .then(() => {
            dispatch(updateUser(uid, updates));
          });
      }
      return db
        .collection('users')
        .doc(uid)
        .set({
          avatar: 'http://www.perlworkshop.nl/nlpw2016/img/default-avatar.png',
          email: '',
          username: '',
          stars: 0,
          topCategories: ['life', 'programming', 'Music']
        })
        .then(startUpdateUser(uid, updates));
    });
};

export const addUserStar = nextStars => ({
  type: 'ADD_USER_STAR',
  nextStars
});

export const startAddUserStar = (id, prevStars) => dispatch =>
  db
    .collection('users')
    .doc(id)
    .update({ stars: prevStars + 1 })
    .then(() => dispatch(addUserStar(prevStars + 1)));

export const setAuthor = author => ({
  type: 'SET_AUTHOR',
  author
});

export const startSetAuthor = id => dispatch =>
  db
    .collection('users')
    .doc(id)
    .get()
    .then((doc) => {
      dispatch(setAuthor(doc.data()));
    });

export const setUserPreferences = preferences => ({
  type: 'SET_USER_PREFERENCES',
  preferences
});

export const startSetUserPreferences = preferences => (dispatch, getState) =>
  db
    .collection('users')
    .doc(getState().auth.uid)
    .get()
    .then((doc) => {
      dispatch(setUserPreferences(doc.data()));
    });
