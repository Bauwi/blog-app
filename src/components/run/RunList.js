import React, { Component } from 'react';

import RunListItem from './RunListItem';

export default class RunList extends Component {
  renderList() {
    return this.props.posts.map((post) => {
      const key = post.content ? post.content.id : post.id;

      return <RunListItem key={key} post={post} state={post.state} />;
    });
  }

  render() {
    return (
      <div className="content-container run-list">
        <p>RunList</p>
        {this.props.posts ? (
          this.renderList()
        ) : (
          <div>Add a post to your run to start a session</div>
        )}
      </div>
    );
  }
}
