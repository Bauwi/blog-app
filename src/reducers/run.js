const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
