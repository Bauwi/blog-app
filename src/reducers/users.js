const usersReducerDefaultState = [];

const usersReducer = (state = usersReducerDefaultState, action) => {
  switch (action.type) {
    case 'CREATE_USER':
      return {
        ...state,
        preferences: action.newUser
      };
    case 'UPDATE_USER_PREFERENCES':
      return {
        ...state,
        preferences: { ...state.preferences, ...action.updates }
      };
    case 'SET_AUTHOR':
      return {
        ...state,
        author: action.author
      };
    case 'ADD_USER_STAR':
      return {
        ...state,
        author: {
          ...state.author,
          stars: action.nextStars
        }
      };
    case 'SET_USER_PREFERENCES':
      return {
        ...state,
        preferences: action.preferences
      };
    case 'USERS_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'USERS_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    default:
      return state;
  }
};

export default usersReducer;
