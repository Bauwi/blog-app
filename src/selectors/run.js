const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
