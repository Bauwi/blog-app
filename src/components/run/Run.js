import React, { Component } from 'react';
import { connect } from 'react-redux';

import InlineList from '../InlineList';
import LoadingPage from '../LoadingPage';
import { startSetRunPosts } from '../../actions/run';

export class Run extends Component {
  state = {
    loading: true
  };
  componentWillMount() {
    if (this.props.run.length === 0) {
      this.props.startSetRunPosts().then(() => {
        this.setState(() => ({ loading: false }));
      });
    } else {
      this.setState(() => ({ loading: false }));
    }
  }

  render() {
    if (this.state.loading) {
      return <LoadingPage />;
    }
    return (
      <div>
        <p>Run</p>
        <InlineList posts={this.props.run} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startSetRunPosts: () => dispatch(startSetRunPosts())
});

const mapStateToProps = state => ({
  run: state.run
});

export default connect(mapStateToProps, mapDispatchToProps)(Run);
