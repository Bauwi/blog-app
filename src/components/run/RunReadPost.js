import React, { Component } from 'react';

import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import Img from 'react-image';

import UserCard from '../UserCard';
import LoadingPage from '../LoadingPage';
import RunReadPostEditor from './RunReadPostEditor';

import { startSetAuthorFromUserId, startAddUserStar } from '../../actions/users';
import { startUpPostStar } from '../../actions/posts';

export class RunReadPost extends Component {
  state = {
    loading: true
  };

  componentWillMount() {
    const { id } = this.props.post.content;
    this.props.startSetAuthorFromUserId(this.props.post.content.authorId).then(() => {
      this.setState(() => ({
        loading: false
      }));
    });
  }

  onAddStar = () => {
    this.props.startAddUserStar(this.props.post.content.authorId, this.props.author.stars);
    this.props.startUpPostStar(this.props.post.content.id);
  };

  render() {
    if (this.state.loading) {
      return <LoadingPage />;
    }
    const Cover = () => (
      <Img
        className="image image--cover"
        src={cover}
        loader={
          <div className="loader">
            <img className="loader__image" src="/images/loader.gif" alt="Loading..." />
          </div>
        }
      />
    );
    const { title, body, keywords, cover, stars } = this.props.post.content;
    const delta = { ops: [...body] };
    return (
      <div>
        <header className="page-header page-header--read">
          <div className="content-container">
            <UserCard post={this.props.post.content} author={this.props.author} />{' '}
            <div className="read-header__stars">
              <span>{stars}</span>
              <button className="button button--star" onClick={this.onAddStar}>
                <i className="fa fa-star-o" />
              </button>
              <p> {this.props.postStars}</p>{' '}
            </div>{' '}
          </div>{' '}
        </header>
        <h2 className="read-header__title"> {title} </h2>
        <Cover />
        <RunReadPostEditor delta={delta} />

        <footer> keywords: {keywords} </footer>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startSetAuthorFromUserId: id => dispatch(startSetAuthorFromUserId(id)),
  startAddUserStar: (id, prevStars) => dispatch(startAddUserStar(id, prevStars)),
  startUpPostStar: id => dispatch(startUpPostStar(id))
});

const mapStateToProps = (state, props) => ({
  author: state.users.author
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(RunReadPost));
