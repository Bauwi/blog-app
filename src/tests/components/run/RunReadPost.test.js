import React from 'react';
import { shallow } from 'enzyme';
import { RunReadPost } from '../../../components/run/RunReadPost';
import run from '../../fixtures/run';
import { user2 } from '../../fixtures/users';

test('should render RunReadPost component properly', () => {
  const wrapper = shallow(<RunReadPost
    isLoading={false}
    post={run.posts[0]}
    startSetAuthorFromUserId={() => {}}
    startUpPostStar={() => {}}
  />);
  expect(wrapper).toMatchSnapshot();
});

test('should render RunReadPost component properly on loading', () => {
  const wrapper = shallow(<RunReadPost
    isLoading
    post={run.posts[0]}
    startSetAuthorFromUserId={() => {}}
    startUpPostStar={() => {}}
  />);
  expect(wrapper).toMatchSnapshot();
});

test('should call startSetAuthorFromUserId on mount', () => {
  const startSetAuthorFromUserId = jest.fn();
  const wrapper = shallow(<RunReadPost
    isLoading
    post={run.posts[0]}
    startSetAuthorFromUserId={startSetAuthorFromUserId}
    startUpPostStar={() => {}}
  />);
  expect(startSetAuthorFromUserId).toHaveBeenLastCalledWith(run.posts[0].content.authorId);
});

test('should call startSetAuthorFromUserId on update', () => {
  const startSetAuthorFromUserId = jest.fn();
  const wrapper = shallow(<RunReadPost
    isLoading
    post={run.posts[0]}
    startSetAuthorFromUserId={startSetAuthorFromUserId}
    startUpPostStar={() => {}}
  />);
  wrapper.setProps({ post: run.posts[1] });
  expect(startSetAuthorFromUserId).toHaveBeenLastCalledWith(run.posts[1].content.authorId);
});

test('should call startUpPostStar on star click', () => {
  const startUpPostStar = jest.fn();
  const wrapper = shallow(<RunReadPost
    isLoading
    author={user2}
    post={run.posts[0]}
    startSetAuthorFromUserId={() => {}}
    startUpPostStar={startUpPostStar}
  />);
  setTimeout(() => {
    wrapper
      .find('button')
      .at(0)
      .simulate('click');
    expect(startUpPostStar).toHaveBeenLastCalledWith(
      run.posts[0].content.id,
      run.posts[0].content.authorId,
      user2.stars
    );
  }, 1000);
});
