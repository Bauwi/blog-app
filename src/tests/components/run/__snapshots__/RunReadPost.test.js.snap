// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RunReadPost component properly 1`] = `
<div>
  <header
    className="page-header page-header--read"
  >
    <div
      className="content-container"
    >
      <UserCard
        author={
          Object {
            "avatar": "https://www.timeshighereducation.com/sites/default/files/byline_photos/default-avatar.png",
            "description": "I had three chairs in my house; one for solitude, two for friendship, three for society.",
            "stars": 0,
            "username": "Anonymous",
          }
        }
        post={
          Object {
            "DBid": "DBid1",
            "content": Object {
              "author": "author1",
              "authorId": "author1Id",
              "body": Array [
                Object {
                  "attributes": "h1",
                  "text": "This is the first Post in run",
                },
              ],
              "cover": "https://www.w3schools.com/w3css/img_fjords.jpg",
              "createdAt": 0,
              "id": "1",
              "keywords": "red, green, blue",
              "miniCover": "https://www.w3schools.com/w3css/img_fjords.jpg",
              "readingTime": 7,
              "stars": 5,
              "title": "Post 1 in Run",
            },
            "state": "read",
          }
        }
      />
      <div
        className="read-header__datereadingtime border-top-item-red"
      >
        <p>
          Jan 1st, 1970
        </p>
        <p>
          <i
            className="fa fa-clock-o"
          />
           
          7 minutes read
        </p>
        <div
          className="read-header__stars"
        >
          <button
            className="button button--star"
            onClick={[Function]}
          >
            5
            <i
              className="fa fa-star-o"
            />
          </button>
          <p />
        </div>
      </div>
    </div>
  </header>
  <h2
    className="read-header__title"
  >
     
    Post 1 in Run
     
  </h2>
  <Cover />
  <RunReadPostEditor
    delta={
      Object {
        "ops": Array [
          Object {
            "attributes": "h1",
            "text": "This is the first Post in run",
          },
        ],
      }
    }
  />
  <footer
    className="keywords-list__read-post-container border-top-item-red"
  >
     
    <KeywordsList
      keywords="red, green, blue"
    />
     
  </footer>
</div>
`;

exports[`should render RunReadPost component properly on loading 1`] = `<Component />`;
